### DISCUSSION TITLE
How can I allow node deletion via keyboard with a confirmation prompt?

### DISCUSSION THREAD

**Discussion by AdamGerthel** [xyflow/xyflow#2925] [Q&A] [answered]
I'm using `onNodesDelete` to open up a window.confirm dialogue. When the user accepts the deletion, the node is deleted from my datastore which in turn will update ReactFlow's nodes. However, it seems like ReactFlow will remove the node no matter what. It doesn't affect my application's state, but the behavior is weird because the user will think that they deleted the node even if they selected cancel.

So my question is this: How can I get ReactFlow to *not* remove a node when the delete key is triggered? I want to implement my own delete function, and not have the node automatically be removed from the canvas.

**Comment by AdamGerthel** [comment, marked as answer]
I solved it by adding my custom key handles using `useKeyPress` and disabled the built in `deleteKeyCode` prop by passing it an empty array.

`deleteKeyCode={[]}`

**Reply by cbureriu** [reply]
could you post a working example?

**Reply by AdamGerthel** [reply]
I don't have time to set up a codesandbox or anything like that but I can paste snippets of my code.

Here's how I "listen" to delete presses:

```tsx
const deletePressed = useKeyPress(['Delete', 'Backspace'])

  useEffect(() => {
    // Do something here
  }, [deletePressed])
```

And below is my ReactFlow component, where you can see that I'm passing an empty array to `deleteKeyCode` which prevents it from hi-jacking my `useKeyPress` implementation.

```tsx
      <ReactFlow
        nodes={nodes}
        edges={edges}
        onNodesChange={setNewNodePositionWithThrottle}
        onEdgesChange={onEdgesChange}
        onConnect={onConnect}
        maxZoom={1}
        minZoom={0.1}
        nodeTypes={nodeTypes}
        zoomOnScroll={false}
        fitView
        deleteKeyCode={[]}
      >
        <Controls />
        <MiniMap />
        <Background variant={BackgroundVariant.Dots} gap={12} size={1} />
      </ReactFlow>
```

This is using reactflow `11.5.6`

**Reply by baibhavbista** [reply]
Some more details on top of @AdamGerthel 's solution:

What you generally want to do is 
1. Get the selected nodes & edges: can use the callback [onSelectionChange](https://reactflow.dev/api-reference/react-flow#on-selection-change) on the `<ReactFlow />` component
2. Do the confirmation logic/prompt: this would be your application specific logic
2. Then, delete the nodes and edges: via helper function [reactFlowInstance.deleteElements](https://reactflow.dev/api-reference/types/react-flow-instance#delete-elements)

**Comment by neelduttahere** [comment]
> I don't have time to set up a codesandbox or anything like that but I can paste snippets of my code.
> 
> Here's how I "listen" to delete presses:
> 
> ```tsx
> const deletePressed = useKeyPress(['Delete', 'Backspace'])
> 
>   useEffect(() => {
>     // Do something here
>   }, [deletePressed])
> ```
> 
> And below is my ReactFlow component, where you can see that I'm passing an empty array to `deleteKeyCode` which prevents it from hi-jacking my `useKeyPress` implementation.
> 
> ```tsx
>       <ReactFlow
>         nodes={nodes}
>         edges={edges}
>         onNodesChange={setNewNodePositionWithThrottle}
>         onEdgesChange={onEdgesChange}
>         onConnect={onConnect}
>         maxZoom={1}
>         minZoom={0.1}
>         nodeTypes={nodeTypes}
>         zoomOnScroll={false}
>         fitView
>         deleteKeyCode={[]}
>       >
>         <Controls />
>         <MiniMap />
>         <Background variant={BackgroundVariant.Dots} gap={12} size={1} />
>       </ReactFlow>
> ```
> 
> This is using reactflow `11.5.6`

I was searching for a way to disable deleting nodes and edges for all nodes and edges, using the delete key. Setting `deleteKeyCode={[]}` seems to be the best solution for me! As based on my implementation I am using custom business logic to delete nodes and edges on canvas. Thanks @AdamGerthel !

**Reply by moklick** [reply]
In v12 you will have a `onBeforeDelete` handler that you could use for this :)
